ricecooker/utils/youtube_cache.py

```diff
@@ -0,0 +1,152 @@
+                        For full list of available options: https://github.com/ytdl-org/youtube-dl/blob/master/youtube_dl/YoutubeDL.py
+        :return: A ricecooker-like info dict info about the video or None if extraction fails
+        """
+        # 1. Try to get from cache if allowed:
+        vinfo = None
+        if use_cache and os.path.exists(self.vinfo_json_path):
+            vinfo = json.load(open(self.vinfo_json_path))
+            LOGGER.info("==> [Video %s] Retrieving cached video information...", self.cachename)
+        # 2. Fetch info from youtube_dl
+        if not vinfo:
+            LOGGER.info("==> [Video %s] Requesting %s from youtube...", self.cachename, self.url)
+            try:
+                video = YouTubeResource(self.url)
+            except youtube_dl.utils.ExtractorError as e:
+                if "unavailable" in str(e):
+                    LOGGER.error("==> [Video %s] Video not found at URL: %s", self.cachename, self.url)
+                    return None
+
+            if video:
+                try:
+                    # Save video info to JSON cache file
+                    vinfo = video.get_resource_info(options)
+                    json.dump(vinfo,
+                              open(self.vinfo_json_path, 'w'),
+                              indent=4,
+                              ensure_ascii=False,
+                              sort_keys=True)
+                except Exception as e:
+                    LOGGER.error("==> [Video %s] Failed to get video info: %s", self.cachename, e)
+                    return None
+            else:
+                return None
+        return vinfo
+
+
+class YouTubePlaylistCache(object):
+
+    def __init__(self, playlist_id, alias=''):
+        """
+        Initializes YouTubePlaylistCache object
+        :param alias: Alias name for the JSON cache filename, which will be named as playlist_id if such field not specified
+        """
+        self.playlist_id = playlist_id
+        if not alias:
+            self.cachename = alias
+        else:
+            self.cachename = self.playlist_id
+        if not os.path.isdir(YOUTUBE_CACHE_DIR):
+            os.mkdir(YOUTUBE_CACHE_DIR)
+        self.playlist_info_json_path = os.path.join(YOUTUBE_CACHE_DIR, self.cachename + '.json')
+
+    def __str__(self):
+        return 'YouTubePlaylistCache (%s)' % (self.cachename)
+
+    def get_playlist_info(self, use_cache=True, youtube_ignore_error=True, youtube_skip_download=True, options=None):
+        """
+        Get YouTube playlist info by either requesting URL or extracting local cache
+        :param use_cache: Define if allowed to get playlist info from local JSON cache, default to True
+        :param youtube_ignore_error: Do not stop on download errors.
+                                     Please enable this option when videos of playlist is private or deleted thus extraction won't be blocked on those videos
```

Comment:
Setting `youtube_ignore_error=True` is known to cause problems with playlists when using the web proxy. We need to raise errors in cases where a network error occurs 429 (cheffing server IP blocked by youtube), so that we change to a different proxy server, but if we give `ignoreerrors=True` to youtube_dl these errors get absorbed.

On the other hand setting `ignoreerrors=False` breaks the json post-processing logic in YouTubeResource, since it will assume all results are valid so we'll either need to add special handling there (or better create a special class for handling playlist like you have started).

I have another idea for workaround to use [`extract_flat` option](https://github.com/ytdl-org/youtube-dl/blob/master/youtube_dl/YoutubeDL.py#L226) to get just the URLs of playlists resource, then manually download each video (skipping the ones with missing permission and broken ones). This will allow us to reuse all the caching logic for individual videos instead of having a giant json blob for the whole playlist. I will provide more details on this issue https://github.com/learningequality/pressurecooker/issues/32 .

Reply:
Hi Ivan, thanks for your comments. That's a very interesting point on option `ignore_error` that I'm not aware of. The reason I enable this option is that during extraction YouTube info, some of the deleted, private and even normal videos from the playlist will block progress of the whole playlist. So I choose to use this option and meanwhile design a helper function to extract YouTube info for single video and insert that to playlist cache file(for somehow, single YT video extraction is much stable than a YT playlist). Seems more researches and trails are necessary for those options and I will update this code part.